File: pkg/earthsci-ast-go/pkg/esm/join_namespacing_test.go
package esm

// CROSS-BINDING conformance for §5.5.6 join-name namespacing under flattening.
//
// A relational `join` carries its references as plain STRINGS — an `on` key
// column, and an `overlap` clause's `src_env` / `tgt_env` envelope factors —
// inside untyped clause maps. mapExprChildren walks the clause slice, but a
// clause has no "op" key so asExprNode declines it and every name inside used
// to pass through unchanged, leaving a flattened tree whose join named
// variables the flattened registry no longer holds.
//
// The gate is `varNames`, the same one a bare string leaf gets: a key column
// naming a loop symbol or a document-scoped index set is not a declared
// variable of this system and is left alone.
//
// Go's flattened EQUATIONS render to strings, so the surface this reaches is
// events — which is also why it is exercised through namespaceExpressionTree
// directly rather than through Flatten.

import (
	"reflect"
	"testing"
)

func aggregateWithJoin(join []any) Expression {
	return map[string]any{
		"op":         "aggregate",
		"semiring":   "sum_product",
		"output_idx": []any{},
		"ranges": map[string]any{
			"src": map[string]any{"from": "sourceType"},
			"c":   map[string]any{"from": "pop_cells"},
		},
		"join": join,
		"args": []any{"X", "W"},
		"expr": map[string]any{"op": "index", "args": []any{"W", "c"}},
	}
}

func joinOf(t *testing.T, out Expression) []any {
	t.Helper()
	node, ok := asExprNode(out)
	if !ok {
		t.Fatalf("namespaced result is not an expression node: %#v", out)
	}
	return node.Join
}

// The envelope factors of an `overlap` clause name variables, so they follow
// the registry — and so do the same names carried as expression children in
// `args`. A node must not spell one factor two ways.
func TestNamespaceJoin_OverlapEnvelopeFactorsFollowTheRegistry(t *testing.T) {
	varNames := map[string]bool{"X": true, "Y": true, "W": true, "S": true, "E": true, "N": true}
	expr := aggregateWithJoin([]any{
		map[string]any{"overlap": map[string]any{
			"src_env": []any{"X", "Y"},
			"tgt_env": []any{"W", "S", "E", "N"},
			"eps":     0.0,
		}},
	})

	out := namespaceExpressionTree(expr, "ISRM", varNames)
	node, _ := asExprNode(out)

	overlap := joinOf(t, out)[0].(map[string]any)["overlap"].(map[string]any)
	if got, want := overlap["src_env"], []any{"ISRM.X", "ISRM.Y"}; !reflect.DeepEqual(got, want) {
		t.Errorf("src_env = %#v, want %#v", got, want)
	}
	want := []any{"ISRM.W", "ISRM.S", "ISRM.E", "ISRM.N"}
	if got := overlap["tgt_env"]; !reflect.DeepEqual(got, want) {
		t.Errorf("tgt_env = %#v, want %#v", got, want)
	}
	// `eps` is a scalar, not a name: this pass must not touch it. Pinned against
	// the same walk with an empty declared-variable set, so the assertion is
	// about THIS pass and not about how the node decoder spells a number.
	untouched := joinOf(t, namespaceExpressionTree(expr, "ISRM", map[string]bool{}))
	inert := untouched[0].(map[string]any)["overlap"].(map[string]any)
	if !reflect.DeepEqual(overlap["eps"], inert["eps"]) {
		t.Errorf("eps = %#v, want it untouched (%#v)", overlap["eps"], inert["eps"])
	}
	if got, want := inert["src_env"], []any{"X", "Y"}; !reflect.DeepEqual(got, want) {
		t.Errorf("with no declared locals, src_env must be inert: %#v", got)
	}
	// The same factors as expression children, in the same node.
	if got, want := node.Args, []any{"ISRM.X", "ISRM.W"}; !reflect.DeepEqual(got, want) {
		t.Errorf("args = %#v, want %#v", got, want)
	}
}

// The gate is the declared-variable set, not "every bare string": a key column
// naming a loop symbol (`src`, a `ranges` key) or a document-scoped index set
// (`sourceType`) must survive untouched.
func TestNamespaceJoin_LoopSymbolsAndIndexSetsAreLeftAlone(t *testing.T) {
	varNames := map[string]bool{"src_bin": true, "tgt_bin": true, "W": true, "X": true}
	expr := aggregateWithJoin([]any{
		map[string]any{"on": []any{
			[]any{"src_bin", "tgt_bin"},
			[]any{"src", "sourceType"},
		}},
	})

	on := joinOf(t, namespaceExpressionTree(expr, "M", varNames))[0].(map[string]any)["on"]
	want := []any{
		[]any{"M.src_bin", "M.tgt_bin"},
		[]any{"src", "sourceType"},
	}
	if !reflect.DeepEqual(on, want) {
		t.Errorf("on = %#v, want %#v", on, want)
	}
}

// A join-free node is byte-identical, and an unrecognised clause shape is
// carried through rather than dropped.
func TestNamespaceJoin_UnknownClauseShapeIsPreserved(t *testing.T) {
	varNames := map[string]bool{"W": true, "X": true}
	expr := aggregateWithJoin([]any{"not-a-clause"})
	if got := joinOf(t, namespaceExpressionTree(expr, "M", varNames)); !reflect.DeepEqual(got, []any{"not-a-clause"}) {
		t.Errorf("unrecognised clause = %#v, want it preserved", got)
	}
}

// A LOOP SYMBOL SHADOWED by a same-named declared variable stays a loop symbol.
// esm-spec §4.3.1 lets one string be a variable reference in most contexts and
// an index symbol inside an aggregate's `output_idx` / `expr` / `ranges` keys, so
// a system may legally declare a variable named `src` while an aggregate binds
// `src` as a range. Inside the node the string denotes the LOOP SYMBOL, and an
// `on` key column is resolved against this node's own ranges — so prefixing it
// makes it resolve to nothing. The binder set must therefore win over the
// declared-variable gate.
func TestNamespaceJoin_ShadowedLoopSymbolStaysALoopSymbol(t *testing.T) {
	// `src` is BOTH a declared variable and a `ranges` key of the node.
	varNames := map[string]bool{"src": true, "src_bin": true, "tgt_bin": true, "W": true, "X": true}
	expr := aggregateWithJoin([]any{
		map[string]any{"on": []any{
			[]any{"src", "sourceType"},
			[]any{"src_bin", "tgt_bin"},
		}},
	})

	on := joinOf(t, namespaceExpressionTree(expr, "M", varNames))[0].(map[string]any)["on"]
	want := []any{
		[]any{"src", "sourceType"},
		[]any{"M.src_bin", "M.tgt_bin"},
	}
	if !reflect.DeepEqual(on, want) {
		t.Errorf("on = %#v, want %#v (a shadowed loop symbol must not be prefixed)", on, want)
	}
}

// ...and the same for an `output_idx` entry, the other binder position.
func TestNamespaceJoin_ShadowedOutputIndexStaysABinder(t *testing.T) {
	expr := aggregateWithJoin([]any{
		map[string]any{"on": []any{[]any{"o", "sourceType"}}},
	}).(map[string]any)
	// `o` is an output index AND (below) a declared variable. The literal
	// singleton dimension beside it is what the binder scan must skip rather than
	// choke on — output_idx admits Int 1 entries (esm-spec §4.3.1).
	expr["output_idx"] = []any{"o", 1}

	varNames := map[string]bool{"o": true}
	on := joinOf(t, namespaceExpressionTree(expr, "M", varNames))[0].(map[string]any)["on"]
	if got, want := on, []any{[]any{"o", "sourceType"}}; !reflect.DeepEqual(got, want) {
		t.Errorf("on = %#v, want %#v (an output_idx binder must not be prefixed)", got, want)
	}
}

File: pkg/EarthSciAST.jl/test/join_namespacing_test.jl
# CROSS-BINDING conformance for §5.5.6 join-name namespacing under flattening.
#
# A relational `join` carries its references as plain STRINGS — an `on` key
# column, and an `overlap` clause's `src_env` / `tgt_env` envelope factors —
# rather than as `VarExpr` children. That is an ENCODING choice, not a scoping
# one: the value-invention materialiser resolves every one of those names
# against the variable registry (`_vi_join_index_sym` → `ctx.variables`,
# `_vi_env_vectors` → `ctx.const_arrays`), and after flattening that registry is
# the dot-namespaced one.
#
# Julia has ALWAYS namespaced them (`namespacing.jl::_namespace_join`), so these
# cases pass before and after the cross-binding alignment. They are the
# REFERENCE the Rust `tests/join_namespacing.rs` and Python
# `tests/test_join_namespacing.py` cases were written against — Rust and Python
# both left `join` bare, so the same document flattened by the three bindings
# produced join clauses that disagreed.
#
# The gate is the component's declared LOCAL names: a key column naming a loop
# symbol or a document-scoped index set is not a local variable and is left
# alone, which is what lets one rule serve `on` and `overlap` alike.

module JoinNamespacingTests

using Test
using EarthSciAST
import JSON3

const ESS = EarthSciAST

const REPO_ROOT = normpath(joinpath(@__DIR__, "..", "..", ".."))
const OVERLAP_FIXTURE = joinpath(REPO_ROOT, "pkg", "earthsci-ast-rs", "tests",
                                      "fixtures", "pushdown",
                                      "overlap_gate_point_in_rect.esm")
const JOIN_FILTER_FIXTURE = joinpath(REPO_ROOT, "tests", "valid", "aggregate",
                                     "join_filter.esm")

# The one flattened equation whose RHS carries a `join`.
function producer(flat)
    for eq in flat.equations
        r = eq.rhs
        if r isa ESS.OpExpr && r.join !== nothing && !isempty(r.join)
            return r
        end
    end
    error("no join-bearing equation in the flattened system")
end

@testset "join namespacing under flattening (§5.5.6)" begin

    @testset "overlap envelope factors follow the variable registry" begin
        flat = ESS.flatten(ESS.load(OVERLAP_FIXTURE))
        node = producer(flat)
        clause = node.join[1]
        @test clause isa ESS._OverlapJoinSpec
        @test clause.src_env == ["ISRM.X", "ISRM.Y"]
        @test clause.tgt_env == ["ISRM.W", "ISRM.S", "ISRM.E", "ISRM.N"]

        # The SAME six factors also appear as expression children in `args`, so
        # the two spellings inside one node must agree — the sharpest statement
        # of why these are references and not opaque metadata.
        argnames = sort([a.name for a in node.args if a isa ESS.VarExpr])
        @test argnames == sort(vcat(clause.src_env, clause.tgt_env))

        # ...and that is the registry they must resolve in.
        for n in vcat(clause.src_env, clause.tgt_env)
            @test haskey(flat.parameters, n)
        end
    end

    @testset "the flattened producer materialises the golden support set" begin
        flat = ESS.flatten(ESS.load(OVERLAP_FIXTURE))
        # The ISRM L1 geometry, keyed by the FLATTENED names.
        ca = Dict{String, Any}(
            "ISRM.W" => [0.0, 2.0, 4.0, 0.0, 2.0, 4.0, 0.0, 2.0, 4.0],
            "ISRM.E" => [2.0, 4.0, 6.0, 2.0, 4.0, 6.0, 2.0, 4.0, 6.0],
            "ISRM.S" => [0.0, 0.0, 0.0, 2.0, 2.0, 2.0, 4.0, 4.0, 4.0],
            "ISRM.N" => [2.0, 2.0, 2.0, 4.0, 4.0, 4.0, 6.0, 6.0, 6.0],
            "ISRM.X" => [1.0, 3.0, 1.0, 5.0, 1.5],
            "ISRM.Y" => [1.0, 1.0, 3.0, 5.0, 1.5],
        )
        # Round-trip the flattened system back into a document (the front door
        # the value-invention engine consumes) and materialise it.
        reconstituted = ESS.load(ESS.flattened_to_esm(flat))
        model = first(values(reconstituted.models))
        vi = ESS.materialize_value_invention(model, reconstituted.index_sets, ca,
                                             Dict{String, Float64}())
        @test vi.members["emis_src_cells_faq"] == [1, 2, 4, 9]
    end

    @testset "loop symbols and index sets are NOT namespaced" begin
        # The committed M2 fixture joins two LOOP SYMBOLS (`src`/`fuel`, keys of
        # `ranges`) against two DOCUMENT-scoped index sets. Neither kind is a
        # component-local variable, so the declared-local gate leaves all four
        # alone — the case that makes "namespace every bare string" wrong.
        flat = ESS.flatten(ESS.load(JOIN_FILTER_FIXTURE))
        node = producer(flat)
        @test node.join[1] == [("src", "sourceType"), ("fuel", "fuelType")]
    end

    @testset "a key column naming a local buffer IS namespaced" begin
        d = JSON3.read(read(JOIN_FILTER_FIXTURE, String), Dict{String, Any})
        m = d["models"]["EmissionsAggregate"]
        m["variables"]["src_bin"] = Dict("type" => "parameter")
        m["variables"]["tgt_bin"] = Dict("type" => "parameter")
        m["equations"][1]["rhs"]["join"] = Any[Dict("on" => Any[
            Any["src_bin", "tgt_bin"], Any["src", "sourceType"]])]
        node = producer(ESS.flatten(ESS.load(d)))
        @test node.join[1] == [
            ("EmissionsAggregate.src_bin", "EmissionsAggregate.tgt_bin"),
            ("src", "sourceType"),
        ]
    end

    @testset "variable_map removal carries the join names with it" begin
        # `param_to_var` REMOVES the mapped parameter, so a join still naming it
        # points at a variable the flattened system no longer declares
        # (`_rename_join_names`).
        d = JSON3.read(read(OVERLAP_FIXTURE, String), Dict{String, Any})
        d["models"]["EDGE"] = Dict(
            "variables" => Dict("src_W" => Dict("type" => "parameter",
                                                "shape" => Any["pop_cells"])),
            "equations" => Any[],
        )
        d["coupling"] = Any[Dict("type" => "variable_map",
                                 "from" => "EDGE.src_W",
                                 "to" => "ISRM.W",
                                 "transform" => "param_to_var")]
        flat = ESS.flatten(ESS.load(d))
        @test !haskey(flat.parameters, "ISRM.W")
        clause = producer(flat).join[1]
        @test clause.tgt_env[1] == "EDGE.src_W"
    end

end

end # module
